File: oauth2-server/src/main/java/com/clouway/oauth2/UserInfoController.java
package com.clouway.oauth2;

import com.clouway.friendlyserve.Request;
import com.clouway.friendlyserve.Response;
import com.clouway.friendlyserve.RsJson;
import com.clouway.oauth2.token.Token;
import com.clouway.oauth2.token.Tokens;
import com.clouway.oauth2.user.IdentityFinder;
import com.google.common.base.Optional;
import com.google.gson.JsonObject;

import java.util.Map;

/**
 * @author Miroslav Genov (miroslav.genov@clouway.com)
 */
class UserInfoController implements InstantaneousRequest {

  private final IdentityFinder identityFinder;
  private final Tokens tokens;

  UserInfoController(IdentityFinder identityFinder, Tokens tokens) {
    this.identityFinder = identityFinder;
    this.tokens = tokens;
  }

  @Override
  public Response handleAsOf(Request request, DateTime instantTime) {
    String accessToken = request.param("access_token");

    Optional<Token> possibleTokenResponse = tokens.findTokenAvailableAt(accessToken, instantTime);

    if (!possibleTokenResponse.isPresent()) {
      return OAuthError.invalidToken();
    }

    Token token = possibleTokenResponse.get();

    Optional<Identity> possibleIdentityResponse = identityFinder.findIdentity(token.identityId, token.grantType, instantTime);
    if (!possibleIdentityResponse.isPresent()) {
      return OAuthError.unknownClient();
    }

    Identity identity = possibleIdentityResponse.get();

    JsonObject o = new JsonObject();

    o.addProperty("id", identity.id());
    o.addProperty("name", identity.name());
    o.addProperty("email", identity.email());
    o.addProperty("given_name", identity.givenName());
    o.addProperty("family_name", identity.familyName());

    Map<String, Object> claims = identity.claims();

    for (String key : claims.keySet()) {
      Object value = claims.get(key);

      if (value instanceof String) {
        o.addProperty(key, (String) value);
      }
      if (value instanceof Number) {
        o.addProperty(key, (Number) value);
      }

      if (value instanceof Boolean) {
        o.addProperty(key, (Boolean) value);
      }
    }

    return new RsJson(o);
  }
}

File: oauth2-server/src/main/java/com/clouway/oauth2/UserInfoController.kt
package com.clouway.oauth2

import com.clouway.friendlyserve.Request
import com.clouway.friendlyserve.Response
import com.clouway.friendlyserve.RsJson
import com.clouway.oauth2.common.DateTime
import com.clouway.oauth2.token.FindIdentityRequest
import com.clouway.oauth2.token.FindIdentityResult
import com.clouway.oauth2.token.IdentityFinder
import com.clouway.oauth2.token.Tokens
import com.google.gson.JsonObject

/**
 * @author Miroslav Genov (miroslav.genov@clouway.com)
 */
internal class UserInfoController(
    private val identityFinder: IdentityFinder,
    private val tokens: Tokens,
) : InstantaneousRequest {
    override fun handleAsOf(
        request: Request,
        instantTime: DateTime,
    ): Response {
        val accessToken = request.param("access_token")
		
        val possibleTokenResponse = tokens.findTokenAvailableAt(accessToken, instantTime)
		
        if (!possibleTokenResponse.isPresent) {
            return OAuthError.invalidToken("Access token was not found.")
        }
		
        val token = possibleTokenResponse.get()
		
        val findReq = FindIdentityRequest(
            subject = token.subject,
            instantTime = instantTime,
            clientId = token.clientId,
        )
		
        val o = JsonObject()
        var claims: Map<String, Any> = mutableMapOf()
        when (val res = identityFinder.findIdentity(findReq)) {
            is FindIdentityResult.User -> {
                val identity = res.identity
                o.addProperty("id", identity.id)
                o.addProperty("name", identity.name)
                o.addProperty("email", identity.email)
                o.addProperty("given_name", identity.givenName)
                o.addProperty("family_name", identity.familyName)

                claims = identity.claims
            }
            is FindIdentityResult.ServiceAccountClient -> {
                val serviceAccount = res.serviceAccount
                o.addProperty("id", serviceAccount.clientId)
                o.addProperty("name", serviceAccount.name)
                o.addProperty("email", serviceAccount.clientEmail)
                o.addProperty("given_name", "")
                o.addProperty("family_name", "")

                claims = serviceAccount.claims
            }
            is FindIdentityResult.NotFound -> {
                return OAuthError.invalidGrant("Identity was not found.")
            }
            is FindIdentityResult.Error -> {
                return OAuthError.internalError()
            }
        }

        for (key in claims.keys) {
            val value = claims[key]
			
            if (value is String) {
                o.addProperty(key, value)
            }
            if (value is Number) {
                o.addProperty(key, value)
            }
			
            if (value is Boolean) {
                o.addProperty(key, value)
            }
        }
		
        return RsJson(o)
    }
}
